src/Sulu/Bundle/CategoryBundle/Trash/CategoryTrashItemHandler.php

```diff
@@ -0,0 +1,234 @@
+    }
+
+    /**
+     * @param CategoryInterface $category
+     */
+    public function store(object $category): TrashItemInterface
+    {
+        Assert::isInstanceOf($category, CategoryInterface::class);
+
+        $data = [
+            'key' => $category->getKey(),
+            'defaultLocale' => $category->getDefaultLocale(),
+            'parentId' => $category->getParent() ?  $category->getParent()->getId() : null,
+            'created' => $category->getCreated()->format('c'),
+            'creatorId' => $category->getCreator() ? $category->getCreator()->getId() : null,
+            'metas' => [],
+            'translations' => [],
+        ];
+
+        /** @var CategoryMetaInterface $meta */
+        foreach ($category->getMeta() as $meta) {
+            $metaData = [
+                'key' => $meta->getKey(),
+                'value' => $meta->getValue(),
+                'locale' => $meta->getLocale(),
+            ];
+
+            $data['metas'][] = $metaData;
+        }
+
+        /** @var CategoryTranslationInterface $translation */
+        foreach ($category->getTranslations() as $translation) {
+            $translationData = [
+                'translation' => $translation->getTranslation(),
+                'description' => $translation->getDescription(),
+                'locale' => $translation->getLocale(),
+                'created' => $translation->getCreated()->format('c'),
+                'creatorId' => $translation->getCreator() ? $translation->getCreator()->getId() : null,
+                'mediaIds' => [],
+                'keywords' => [],
+            ];
+
+            /** @var MediaInterface $media */
+            foreach ($translation->getMedias() as $media) {
+                $translationData['mediaIds'][] = $media->getId();
+            }
+
+            /** @var KeywordInterface $keyword */
+            foreach ($translation->getKeywords() as $keyword) {
+                $translationData['keywords'][] = [
+                    'keyword' => $keyword->getKeyword(),
+                    'created' => $keyword->getCreated()->format('c'),
+                    'creatorId' => $keyword->getCreator() ? $keyword->getCreator()->getId() : null,
+                ];
+            }
+
+            $data['translations'][] = $translationData;
+        }
+
+        $categoryTitles = [];
+        foreach ($category->getTranslations() as $translation) {
+            $categoryTitles[$translation->getLocale()] = $translation->getTranslation();
+        }
+
+        return $this->trashItemRepository->create(
+            CategoryInterface::RESOURCE_KEY,
+            (string) $category->getId(),
+            $data,
+            $categoryTitles,
+            CategoryAdmin::SECURITY_CONTEXT,
+            null,
+            null
+        );
+    }
+
+    public function restore(TrashItemInterface $trashItem, array $restoreFormData): object
+    {
+        $data = $trashItem->getRestoreData();
```

Comment:
i thought about moving this logic into the `CategoryManager` (similar to the `TagManager::restore` method). but that way the `CategoryManager` would need to know about the structure of the data returned by `$trashItem->getRestoreData()`. keeping the restore logic here means that the `CategoryTrashItemHandler` is the only class that needs to know the structure of this data.

Reply:
I would keep it this way and not move it to the managers. So all about the trash is in this service and only the store method need to be called in the remove method. Also the CategoryTrashItemHandler should just not be registered when Trash is available. And in future the RemoveHandler should also just call the store method.

Reply:
sounds good to me 🙂  i will remove the `TagManager::restore` too, i dont think that it is a method should be used outside of the `TagTrashItemHandler `